rfcs/20240311-gather-scatter-batching-dims.md

```diff
@@ -0,0 +1,456 @@
+//                   [[0, 1], [1, 1], [0, 9]]
+//                  ],
+//                  [
+//                   [[0, 0], [2, 1], [2, 2]],
+//                   [[1, 2], [0, 1], [1, 0]]
+//                  ]
+//                 ]
+%result = "stablehlo.gather"(%operand, %start_indices) {
+  dimension_numbers = #stablehlo.gather<
+    offset_dims = [3, 4],
+    collapsed_slice_dims = [1],
+    operand_batching_dims = [0],
+    start_indices_batching_dims = [1],
+    start_index_map = [2, 1],
+    index_vector_dim = 3>,
+  slice_sizes = array<i64: 1, 1, 2, 2>,
+  indices_are_sorted = false
+} : (tensor<2x3x4x2xi32>, tensor<2x2x3x2xi64>) -> tensor<2x2x3x2x2xi32>
+// %result: [
+//           [
+//            [
+//             [[1, 2], [3, 4]],
+//             [[3, 4], [5, 6]],
+//             [[13, 14], [15, 16]]
+//            ],
+//            [
+//             [[33, 34], [35, 36]],
+//             [[35, 36], [37, 38]],
+//             [[41, 42], [43, 44]]
+//            ]
+//           ],
+//           [
+//            [
+//             [[1, 2], [3, 4]],
+//             [[13, 14], [15, 16]],
+//             [[21, 22], [23, 24]]
+//            ],
+//            [
+//             [[43, 44], [45, 46]],
+//             [[33, 34], [35, 36]],
+//             [[27, 28], [29, 30]]
+//            ]
+//           ]
+//          ]
+```
+
+&nbsp;[More Examples](../stablehlo/tests/interpret/gather.mlir)
+
+### scatter
+
+#### Semantics
+
+Produces `results` tensors which are equal to `inputs` tensors except that
+several slices specified by `scatter_indices` are updated with the values
+`updates` using `update_computation`.
+
+The following diagram shows how elements in `updates...` map on elements in
+`results...` using a concrete example. The diagram picks a few example
+`updates...` indices and explains in detail which `results...` indices they
+correspond to.
+
+![](images/20240311-gather-scatter-batching-dims/scatter.svg)
+
+More formally, for all `update_index` in `index_space(updates[0])`:
+
+* `update_scatter_dims = [d for d in axes(updates[0]) and d not in
+  update_window_dims]`.
+* `update_scatter_index = update_index[update_scatter_dims...]`.
+* `start_index` is defined as:
+  * `scatter_indices[si0, ..., :, ..., siN]` where `si` are individual
+      elements in `update_scatter_index` and `:` is inserted at the
+      `index_vector_dim` index, if `index_vector_dim` <
+      `rank(scatter_indices)`.
+  * `[scatter_indices[update_scatter_index]]` otherwise.
+* For `d_input` in `axes(inputs[0])`,
+  * `full_start_index[d_input] = start_index[d_start]` if
+    `d_input = scatter_dims_to_operand_dims[d_start]`.
+  * `full_start_index[d_input] = 0` otherwise.
+* For `d_input` in `axes(inputs[0])`,
+  * `full_batching_index[d_input] = update_scatter_index[d_start - (d_start < index_vector_dim ? 0 : 1)]`.
+    if `d_input = operand_batching_dims[i_batching]`, and `d_start = scatter_indices_batching_dims[i_batching]`.
```

Comment:
replace `. if` with `, if`
Also can replace the `, and` -> `and` (there are only two clauses)

Reply:
Done.